extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/JDBCExtractionNamespaceCacheFactory.java

```diff
@@ -76,53 +81,84 @@ public String call()
+        }
+
+        LOG.debug("Updating [%s]", namespace.toString());
+        final List<String[]> columns = dbi.withHandle(
+            new HandleCallback<List<String[]>>()
             {
               @Override
-              public List<Pair<String, String>> withHandle(Handle handle) throws Exception
+              public List<String[]> withHandle(Handle handle) throws Exception
               {
                 final String query;
                 query = String.format(
-                    "SELECT %s, %s FROM %s",
-                    keyColumn,
-                    valueColumn,
+                    makeQueryString(requiredFields),
                     table
                 );
                 return handle
                     .createQuery(
                         query
                     ).map(
-                        new ResultSetMapper<Pair<String, String>>()
+                        new ResultSetMapper<String[]>()
                         {
 
                           @Override
-                          public Pair<String, String> map(
+                          public String[] map(
                               final int index,
                               final ResultSet r,
                               final StatementContext ctx
                           ) throws SQLException
                           {
-                            return new Pair<String, String>(r.getString(keyColumn), r.getString(valueColumn));
+                            String[] values = new String[numFields];
+                            int idx = 0;
+
+                            for (String field: requiredFields)
+                            {
+                              values[idx++] = r.getString(field);
+                            }
+                            return values;
                           }
                         }
                     ).list();
               }
             }
         );
-        for (Pair<String, String> pair : pairs) {
-          cache.put(pair.lhs, pair.rhs);
+        for (String[] values : columns) {
+          for (KeyValueMap keyValueMap: namespace.getMaps())
+          {
+            String mapName = keyValueMap.getMapName();
+            MultiKey key = new MultiKey(id, mapName);
+            Map<String, String> innerMap = cache.get(key);
+            if (innerMap == null)
+            {
+              innerMap = mapAllocator.apply(key);
+              cache.put(key, innerMap);
+            }
+            innerMap.put(values[fieldMap.get(keyValueMap.getKeyColumn())], values[fieldMap.get(keyValueMap.getValueColumn())]);
+          }
         }
         LOG.info("Finished loading %d values for namespace[%s]", cache.size(), id);
         return String.format("%d", System.currentTimeMillis());
       }
     };
   }
 
+  private String makeQueryString(List<String> requiredFields)
+  {
+    String query = "SELECT \"";
+    query += StringUtils.join(requiredFields, "\", \"");
```

Comment:
I don't think this is enough escaping. There could be backslashes and quotes and stuff in the field names. Maybe. Does JDBC/JDBI have a utility function to help with escaping?

EIther that, or let's check the requiredFields against a whitelist of characters.